Create copilot-instructions.md

CarnaViire opened the pull request:
First iteration on the instructions.

Testing on https://github.com/dotnet/runtime/pull/115826.

Despite the remaining environment limitations, the instructions themselves are working.
See e.g. https://github.com/dotnet/runtime/actions/runs/15200538664/job/42753739967#step:9:1582

~~_Note: I've also tried adding the clr build as a [setup step](https://github.com/dotnet/runtime/actions/runs/15200538664/job/42753739967#step:7:3872), but unfortunately it didn't stick_ 🥲~~

CarnaViire committed:
Create copilot-instructions.md

CarnaViire committed:
more instructions

CarnaViire commented:
Based on [the logs](https://github.com/dotnet/runtime/actions/runs/15200159918/job/42752709941#step:8:3165): it used the instructions to build clr runtime
Based on [the same logs](https://github.com/dotnet/runtime/actions/runs/15200159918/job/42752709941#step:8:3359): ...but the build takes too long 🙃

CarnaViire commented on .github/workflows/copilot-setup-steps.yml:
```diff
@@ -17,5 +17,61 @@ jobs:
+          tzdata \
+          unzip \
+          libbpf1 \
+          libelf1t64 \
```
The package list taken from https://github.com/dotnet/dotnet-buildtools-prereqs-docker/blob/4f5930d8206cbd9c53f49882338cdbdfed3fd53b/src/ubuntu/25.04/helix/Dockerfile#L63

CarnaViire commented on .github/copilot-instructions.md:
```diff
@@ -0,0 +1,197 @@
+
+Additionally,
+
+* Don't leave trailling whitespaces.
```
> Comments suppressed due to low confidence (1)
> **.github/copilot-instructions.md:5**
> 
> * There is a typo in the word 'trailling'; it should be 'trailing'.
> 
> ```
> * Don't leave trailling whitespaces.
> ```

```suggestion
* Don't leave trailing whitespaces.
```

CarnaViire committed:
Update .github/copilot-instructions.md

agocke commented on .github/copilot-instructions.md:
```diff
@@ -0,0 +1,197 @@
+1. Checkout `main` branch
+
+2. From the repository root, run the build depending on the affected component. If multiple components are affected, subsequently run and verify the builds for all of them.
+   - **CoreCLR (CLR):** `./build.sh clr+libs`
```
```suggestion
   - **CoreCLR (CLR):** `./build.sh clr+libs+host`
```

The host may or may not be necessary depending on what's going to be run, but this is better safe than sorry.

CarnaViire replied:
I didn't notice it mentioned in the docs (within a "typical scenario") anywhere beside running the Mono runtime tests 😅
Can it be more applicable to runtime, but not that much for the libraries?

I'm just a bit reluctant to add something which is not described the workflow docs as typical and might not be necessary. But I might be missing something, I'm not an expert on building anything beside libraries 😅

CarnaViire replied:
I mean, I've seen that the host exists, but when I open the [building docs](https://github.com/dotnet/runtime/blob/main/docs/workflow/building/coreclr/README.md), that's what I see:

**The Basics**

... main workflow ... build the CoreCLR ...

```
./build.sh -subset clr <other args go here>
```

> _// where are these other args?_

**Build Results**

corerun: ...
coreclr: ...
System.Private.CoreLib.dll: ...

> _// uh-huh_

**What to do with the Build**

> _// uh-huh_

**The Core_Root for Testing Your Build**

```
./build.sh -subset clr+libs -runtimeConfiguration Debug -librariesConfiguration Release
```

> _// cool seems legit_

```
./src/tests/build.sh -x64 -checked -generatelayoutonly
```

> _// ok these are tests already_

**The Dev Shipping Packs**

... full runtime NuGet packages and installer ... more production-esque scenario...

> _// naaah, I don't need that yet_

😅

agocke commented on .github/copilot-instructions.md:
```diff
@@ -0,0 +1,197 @@
+2. From the repository root, run the build depending on the affected component. If multiple components are affected, subsequently run and verify the builds for all of them.
+   - **CoreCLR (CLR):** `./build.sh clr+libs`
+   - **Mono Runtime:** `./build.sh mono+libs`
+   - **Libraries:** `./build.sh clr+libs -rc release`
```
```suggestion
   - **Libraries:** `./build.sh clr+libs+host -rc release`
```

agocke commented on .github/copilot-instructions.md:
```diff
@@ -0,0 +1,197 @@
+From the repository root:
+
+- Build: `./build.sh -subset clr`
+- Run tests: `cd src/tests && ./build.sh`
```
This builds the tests. You need to run `run.sh` to run them. It's mentioned at the bottom of the link below.

CarnaViire replied:
Thanks

agocke commented on .github/copilot-instructions.md:
```diff
@@ -0,0 +1,197 @@
+## 2. Iterative Build and Test Strategy
+
+1. Apply the intended changes
+2. **Attempt Build.** If the build fails, attempt to fix and retry the step (up to 5 attempts).
```
This may work sometimes, but might fail if it tries to build the "wrong" thing. For instance, if it makes a change to the CLR but doesn't rebuild the CLR partition. The libraries build also does deployment, so certain actions may require you to basically rebuild everything.

CarnaViire replied:
> if it makes a change to the CLR but doesn't rebuild the CLR partition

technically this should be covered by the instructions above.
and the agent also does a decent job of trying out different approaches on how to build stuff.

This part is more about actual build failures (like trailing whitespaces 🙃 or, say, the library is targeting netstandard so some of the changes might try to use APIs that are not available)

And in case it goes completely off track, it is asked to limit the number of attempts.

> The libraries build also does deployment, so certain actions may require you to basically rebuild everything.

I don't think I ever needed to do that 🤔 this sounds a bit like a corner case... (or maybe it's just networking libs that aren't affected?)

But in general, I think we don't have to (maybe even _shouldn't_) cover every single possibility here. If we do, we will end up including all the existing workflow docs, and add some more on top. And I'm not sure the result will be helpful for the 90% usecase.

agocke commented on .github/workflows/copilot-setup-steps.yml:
```diff
@@ -17,5 +17,61 @@ jobs:
+          clang \
+          curl \
+          gcc \
+          gdb \
```
This list is probably overkill. I don't think we need `gdb` and I'm almost certain we don't need `gdb` _and_ `lldb`.

The requirements listed in the docs are

```
# requires sudo for non-root user
apt install -y cmake llvm lld clang build-essential \
  python-is-python3 curl git lldb libicu-dev liblttng-ust-dev \
  libssl-dev libkrb5-dev ninja-build pigz cpio
```

CarnaViire replied:
I used the same "better safe than sorry" approach here that you mentioned 😄 but I agree. I'll clean it up.

BTW some of the stuff needs to be added anyway because it's a libmsquic dependency. And bc there is a delay between the new distros coming out, and the packages appearing in the relevant directory in packages.microsoft.com, it is installed manually (and the deps are not brought in)

agocke commented:
> Based on [the same logs](https://github.com/dotnet/runtime/actions/runs/15200159918/job/42752709941#step:8:3359): ...but the build takes too long 🙃 (it takes 16 min and the time limit is 2 min)

This is unattainable. We are nowhere near being able to build any substantial portion of the repo in 2 min.

danmoseley commented:
We likely need to open the firewall too

Chuxel commented:
Re: the setup instructions - generally less is more. For example, rather than spelling out some code formatting rules, you could tell it to run a linter before submitting. That said, if this is giving you the results you are looking for - awesome. Really interesting to see ... also cc @shibbas both for a real-world example and any other tips.

> Based on [the logs](https://github.com/dotnet/runtime/actions/runs/15200159918/job/42752709941#step:8:3165): it used the instructions to build clr runtime Based on [the same logs](https://github.com/dotnet/runtime/actions/runs/15200159918/job/42752709941#step:8:3359): ...but the build takes too long 🙃 (it takes 16 min and the time limit is 2 min)

@ellismg released an update that should resolve the time out in now - let us know if not.

> We likely need to open the firewall too
The good news is we realized the public AzDO package should be open by default since it's a well-known, trusted, public source So that's resolved.  You should be able to ignore `cdn.fwupd.org` unless you literally need Linux firmware updates.  There are a few of these that are getting caught that are things like OS updaters... and its good they're getting caught, but we likely will filter these known things that don't cause harm out.

CarnaViire commented:
@Chuxel 
> Re: the setup instructions - generally less is more. For example, rather than spelling out some code formatting rules, you could tell it to run a linter before submitting. 

The main goal was to teach the agent _how_ to build and test, and how to know _what_ to build and test. TBH at this point the reason I added the code style ones (based on [ASP.NET Core's version](https://github.com/dotnet/aspnetcore/blob/62fb2cece074f30620f2dabc84b97f408f47569b/.github/copilot-instructions.md)) is just to have some additional verification these instructions are actually read. 😅

During the investigation in https://github.com/dotnet/runtime/pull/115826, at some point it felt like it was using some echo of the comment https://github.com/dotnet/runtime/pull/115826#issuecomment-2900918982 rather than copilot-instructions.md I added. And as the comment grew older, I guess that comment was either left out of the context, or severely downweighted, and the agent temporarily forgot how to use `build.sh` 🙃 But by now I see it actually referencing the instructions from the file. Maybe it was confused by the similarity (or some kind of conflict) of the two how-tos.

CarnaViire committed:
Merge branch 'main' into CarnaViire-patch-1

agocke reviewed (approved):
LGTM, thanks!

I think we can adjust as necessary. If it seems like it's struggling, we can add/remove things. The tricky thing with the runtime repo is that everyone has their own workflow because they want to build as little as possible to get to their scenario.

CarnaViire committed:
Code review suggestions, unify formatting

CarnaViire committed:
fix sudo

CarnaViire committed:
Add missing testhost instruction

CarnaViire committed:
Pre-build clr+libs

CarnaViire commented:
I ended up providing the pre-built `clr+libs` to the agent through the setup steps (https://github.com/dotnet/runtime/pull/115917/commits/45b25c6cfc5640bea858e1267221a536d76d58c7), and that helped tremendously.

Before, copilot was still struggling with missing bits and pieces (varying from run to run), and the occasional aggressive build timeouts (but there were cases of successful long builds, too).

Successful run with the latest instructions and setup: [link](https://github.com/dotnet/runtime/actions/runs/15220774998/job/42815855675#step:12:1498)

<details>

```
copilot: Now let's build and run the tests to verify the fix:
<...>
    command: cd /home/runner/work/runtime/runtime && ./build.sh -subset libs -rc release
<...>
    Build succeeded.
        0 Warning(s)
        0 Error(s)
<...>

copilot: Now let's run the specific test to see if our fix resolves the issue:
<...>
    command: cd /home/runner/work/runtime/runtime && dotnet test src/libraries/Microsoft.Extensions.Http/tests/Microsoft.Extensions.Http.Tests/Microsoft.Extensions.Http.Tests.csproj -v n
<...>
   Test Run Successful.
    Total tests: 185
         Passed: 185
     Total time: 9.5574 Seconds
<...>

copilot: Great! The test is now passing. Let's commit this change
```

</details>

I am tempted to :shipit: since I've got an approval from @agocke 
But if someone could pls take a look and sign off the latest addition, that would be great 😄
/cc @ericstj @agocke @danmoseley

CarnaViire commented:
/ba-g copilot infra-only change

ericstj commented on .github/workflows/copilot-setup-steps.yml:
```diff
@@ -17,11 +17,39 @@ jobs:
+          sudo dpkg -i libmsquic* && \
+          rm libmsquic*
+
+      - name: Restore Solution
+        env:
+          RESTORE_ARGS: --restore --excludecibinarylog --warnaserror false /p:BuildAllConfigurations=true /p:DotNetBuildAllRuntimePacks=true
+        # 'DefaultSubsets' don't include 'libs.tests', so restoring it explicitly
+        run: |
+          ./build.sh $RESTORE_ARGS && \
+          ./build.sh libs.tests $RESTORE_ARGS
```
```suggestion
      - name: Restore solution
        run: ./build.sh --restore --excludecibinarylog --warnaserror false /p:BuildAllConfigurations=true /p:DotNetBuildAllRuntimePacks=true /p:DotNetBuildTests=true
```

You can use `DotNetBuildTests`.  We should really have a flag for subsets that means "everything" cc @ViktorHofer

ericstj commented on .github/workflows/copilot-setup-steps.yml:
```diff
@@ -17,11 +17,39 @@ jobs:
+            libxdp1 && \
+          curl -LO "https://packages.microsoft.com/ubuntu/24.04/prod/pool/main/libm/libmsquic/libmsquic_${LIBMSQUIC_VERSION}_amd64.deb" && \
+          sudo dpkg -i libmsquic* && \
+          rm libmsquic*
```
I'd really prefer if we do what was mentioned in https://github.com/dotnet/runtime/issues/115946 to centralize common dependencies and improve the contributor experience too.  Can we create a script `./eng/install-dependencies.sh` that does everything here?

CarnaViire replied:
I totally agree, but the msquic part is tricky. Let's tackle that in a follow-up

ericstj commented on .github/copilot-instructions.md:
```diff
@@ -0,0 +1,249 @@
+    - If the build fails with an error "The shared framework must be built before the local targeting pack can be consumed.", build the runtime (clr or mono) first.
+      E.g., from the repo root, run `./build.sh clr -rc release`.
+
+- **Testhost Is Missing**
+
+    - If a test run fails with errors indicating a missing testhost, such as:
+        - "Failed to launch testhost with error: System.IO.FileNotFoundException", or
+        - "artifacts/bin/testhost/... No such file or directory",
+      this means the required runtime was not built.
+
+    - To resolve, build the appropriate runtime before running tests.
+      For example, run `./build.sh clr -rc release` if testing libraries.
```
These don't look right to me.  Building the testhost is both clr and libs, in particular https://github.com/dotnet/runtime/blob/main/src/libraries/pretest.proj

Maybe we just remove this now that you're pre-building them?

CarnaViire replied:
The case for this failure was when the agent used `./build.sh libs -rc release`, i.e. skipped `clr` from `clr+libs`. That's why I assumed that bringing the missing part should help. But if the `a+b` combination actually brings something additional over the two things `a` and `b` done separately, then yeah, it wouldn't work, but I must say it's a bit counter-intuitive 😅

I worry a bit that removing it completely will bring problems with a similar case but on a mono runtime, but if it does, we can address it separately.

ericstj commented on .github/copilot-instructions.md:
```diff
@@ -0,0 +1,249 @@
+
+        - *Adjust path as needed. If in doubt, search with `find tests -name '*.csproj'`.*
+
+    - `dotnet build /t:test` is generally preferred over `dotnet test`
```
This was fixed last year, but I checked and it does appear to have crept back in  https://github.com/dotnet/runtime/issues/94183

ericstj reviewed (commented):
A few small suggestions.  This looks pretty good.  Thank you for putting together, @CarnaViire

CarnaViire commented on .github/copilot-instructions.md:
```diff
@@ -0,0 +1,249 @@
+
+- **Shared Framework Missing**
+
+    - If the build fails with an error "The shared framework must be built before the local targeting pack can be consumed.", build the runtime (clr or mono) first.
+      E.g., from the repo root, run `./build.sh clr -rc release`.
```
```suggestion
    - If the build fails with an error "The shared framework must be built before the local targeting pack can be consumed.", build both the runtime (clr or mono) and the libs.
      E.g., from the repo root, run `./build.sh clr+libs -rc release` if working on Libraries on CoreCLR. Refer to the section `1.2. Baseline Setup`.
```

CarnaViire committed:
Apply suggestions from code review

Co-authored-by: Eric StJohn <ericstj@microsoft.com>

CarnaViire commented:
Addressed all remaining nits; ready for merge.
Follow-ups tracked in https://github.com/dotnet/runtime/issues/115946.
Time to :shipit:
